### PR TITLE
Allow DA recorded blocks to come out-of-order

### PR DESCRIPTION
**Pull request by MitchTurner**
## Linked Issues/PRs
Closes: https://github.com/FuelLabs/fuel-core/issues/2414

Eliminates the need for: https://github.com/FuelLabs/fuel-core/issues/2397

## Description
It sounds like it is likely we will occasionally receive recorded block information from the committer out of order. There was no real reason to expect them in order so I've generalized the algorithm to accept whatever order.

## Checklist
- [x] New behavior is reflected in tests

### Before requesting review
- [x] I have reviewed the code myself

### CODE REVIEW

File: crates/fuel-gas-price-algorithm/src/v1.rs
Comment (rymnc) [inline crates/fuel-gas-price-algorithm/src/v1.rs:530]: perhaps 
```suggestion
        heights: &[u32],
```
?

Comment (MitchTurner) [reply crates/fuel-gas-price-algorithm/src/v1.rs:530]: Nice catch.

File: crates/fuel-gas-price-algorithm/gas-price-analysis/src/charts.rs
Comment (netrome) [inline crates/fuel-gas-price-algorithm/gas-price-analysis/src/charts.rs:271]: While unrelated to this PR, can we use `i28::from` here instead of an `as` conversion to show that the conversion is infallible (which I hope it is 😅)

File: crates/fuel-gas-price-algorithm/src/v1.rs
Comment (netrome) [inline crates/fuel-gas-price-algorithm/src/v1.rs:528]: Since we no longer take a range here we should rename this function. Perhaps just simplifying the name to
```suggestion
    fn drain_l2_block_bytes(&mut self, heights: &[u32]) -> Result<u128, Error> {
```

File: crates/fuel-gas-price-algorithm/src/v1.rs
Comment (netrome) [inline crates/fuel-gas-price-algorithm/src/v1.rs:546]: Again unrelated but it would be nice to use and explicit infallible conversion: `u128::from`.

File: crates/fuel-gas-price-algorithm/src/v1.rs
Comment (netrome) [inline crates/fuel-gas-price-algorithm/src/v1.rs:303]: It seems like all we do is iterating over the heights, and we do provide a range still in `simulation.rs`. Perhaps we should accept an `impl IntoIterator<Item = u32>` here to allow both ranges and slices to be provided.

No need to do this if you think it's overkill.

File: crates/fuel-gas-price-algorithm/src/v1/tests/update_da_record_data_tests.rs
Comment (netrome) [inline crates/fuel-gas-price-algorithm/src/v1/tests/update_da_record_data_tests.rs:309]: Nit: Not sure if we should call this a range if it's a vec.

```suggestion
    let recorded_heights: Vec<u32> = vec![2, 3];
```

File: crates/fuel-gas-price-algorithm/src/v1/tests/update_da_record_data_tests.rs
Comment (netrome) [inline crates/fuel-gas-price-algorithm/src/v1/tests/update_da_record_data_tests.rs:264]: Same here
```suggestion
    let recorded_heights: Vec<u32> = vec![2, 3];
```

File: crates/fuel-gas-price-algorithm/src/v1/tests/update_da_record_data_tests.rs
Comment (netrome) [inline crates/fuel-gas-price-algorithm/src/v1/tests/update_da_record_data_tests.rs:271]:
```suggestion
    // then
```

File: crates/fuel-gas-price-algorithm/src/v1/tests/update_da_record_data_tests.rs
Comment (netrome) [inline crates/fuel-gas-price-algorithm/src/v1/tests/update_da_record_data_tests.rs:233]: Maybe it's because it's late, but I don't see where we provide blocks out of order in this test case. Is it because we skip `1` in the `recorded_range`?

Comment (netrome) [reply crates/fuel-gas-price-algorithm/src/v1/tests/update_da_record_data_tests.rs:233]: Could we be even more ezplicit and set `recorded_range` to `[3, 2]` instead?.

File: crates/fuel-gas-price-algorithm/src/v1/tests/update_da_record_data_tests.rs
Comment (netrome) [inline crates/fuel-gas-price-algorithm/src/v1/tests/update_da_record_data_tests.rs:308]: The setup here seem identical between this and the previous test. Perhaps we should consider breaking it out to a separate helper function?

**Review by netrome — APPROVED** [review]
Looks good overall to me. A few suggestions, nits and questions.

File: crates/fuel-gas-price-algorithm/gas-price-analysis/src/charts.rs
Comment (MitchTurner) [reply crates/fuel-gas-price-algorithm/gas-price-analysis/src/charts.rs:271]: I've made the change. I think in general the analysis code is a little messier, but now that we're considering keeping it around and maybe expanding it's role it's probably best we get more strict with it.

File: crates/fuel-gas-price-algorithm/src/v1.rs
Comment (MitchTurner) [reply crates/fuel-gas-price-algorithm/src/v1.rs:303]: I think it's overkill. What we'll get from the committer will be a `Vec`. The ranges are just fossils from the old code.

File: crates/fuel-gas-price-algorithm/src/v1/tests/update_da_record_data_tests.rs
Comment (MitchTurner) [reply crates/fuel-gas-price-algorithm/src/v1/tests/update_da_record_data_tests.rs:233]: That's a good idea. I've switched it. In reality, they'll probably be in order inside the vector, but they don't need to be. The point of this PR is that some heights can be skipped and returned to later. These tests skip `1` and do `2` and `3`.

File: crates/fuel-gas-price-algorithm/src/v1/tests/update_da_record_data_tests.rs
Comment (MitchTurner) [reply crates/fuel-gas-price-algorithm/src/v1/tests/update_da_record_data_tests.rs:308]: Yes. I literally had a single test before with 3 asserts, but I broke it out. I have mixed feelings about helpers since they prevent me from accessing whichever values I need from setup.

File: crates/fuel-gas-price-algorithm/src/v1.rs
Comment (rafal-ch) [inline crates/fuel-gas-price-algorithm/src/v1.rs]: Maybe just:

```suggestion
        let recorded_bytes = self.drain_l2_block_bytes(heights)?;
```

It'll save us from renaming the function when we change the datatype again. However, the intention here could be to not relate to the `Range<T>`, but to the "range" as arbitrary "noun".

Either way, just a small detail.

File: crates/fuel-gas-price-algorithm/src/v1.rs
Comment (rafal-ch) [inline crates/fuel-gas-price-algorithm/src/v1.rs:20]: Unrelated, but this could be just `{height}`, not `{height:?}`. But also I see that we use debug for other messages, so maybe there's a reason.

**Review by rafal-ch — APPROVED** [review]
LGTM :+1: 
with just a few nits and minor comments

File: crates/fuel-gas-price-algorithm/gas-price-analysis/src/charts.rs
Comment (netrome) [reply crates/fuel-gas-price-algorithm/gas-price-analysis/src/charts.rs]: Ah right, didn't notice this was in the analysis code :sweat_smile: thanks anyway!

File: crates/fuel-gas-price-algorithm/src/v1.rs
Comment (netrome) [reply crates/fuel-gas-price-algorithm/src/v1.rs:303]: Fair enough :+1:

File: crates/fuel-gas-price-algorithm/src/v1/tests/update_da_record_data_tests.rs
Comment (netrome) [reply crates/fuel-gas-price-algorithm/src/v1/tests/update_da_record_data_tests.rs:308]: > I have mixed feelings about helpers since they prevent me from accessing whichever values I need from setup.

Yeah I can live with a little duplication in test code for this reason. It's easy to make assumptions on shared logic that has to be broken in subsequent tests, so there's no obvious right or wrong here. I'll leave this to your judgement as well.

**Review by netrome — APPROVED** [review]
Thanks for updating! :pray:

File: crates/fuel-gas-price-algorithm/src/v1.rs
Comment (MitchTurner) [reply crates/fuel-gas-price-algorithm/src/v1.rs:550]: This gets called semi-infrequently (only when we get a batch back from the committer). It could become more frequent I guess.

Yes. An alternative would be having the "total unrecorded bytes" tracked and when we remove blocks we subtract their bytes from that total. Then this would just be a multiplication of that. I can add that if you prefer.

File: crates/fuel-gas-price-algorithm/src/v1.rs
Comment (MitchTurner) [reply crates/fuel-gas-price-algorithm/src/v1.rs:538]: It's up to us. If this errors we have a serious problem with our code.

Comment (xgreenx) [reply crates/fuel-gas-price-algorithm/src/v1.rs:538]: If the committer has problems and submits several bundles per same height(they can in theory because they do re-bundle),  it will break the node.

Comment (MitchTurner) [reply crates/fuel-gas-price-algorithm/src/v1.rs:538]: That can't happen if it's finalized before they report to us.

File: crates/fuel-gas-price-algorithm/src/v1.rs
Comment (MitchTurner) [reply crates/fuel-gas-price-algorithm/src/v1.rs]: Done.

File: crates/fuel-gas-price-algorithm/src/v1.rs
Comment (acerone85) [inline crates/fuel-gas-price-algorithm/src/v1.rs:355]: this leaves me with a question. Can we add a block twice, e.g. in the case of a reorganization? If so it might be necessary to subtract the `unrecorded_block_bytes` for the block that is already present at the same height.

**Comment by acerone85** [comment]
had a look, only one question from my side. otherwise LGTM

File: crates/fuel-gas-price-algorithm/src/v1.rs
Comment (MitchTurner) [reply crates/fuel-gas-price-algorithm/src/v1.rs:355]: If the reorg happens on L2, then we will roll back the state of the algorithm too. We regularly store the `GasPriceMetadata` to our database so we can return to previous state. That only happens when we _restart_ the client, it populates the algorithm values with the metadata from that height.

Comment (acerone85) [reply crates/fuel-gas-price-algorithm/src/v1.rs:355]: Thanks, really clear explanation :D

**Review by acerone85 — DISMISSED** [review]

File: crates/fuel-gas-price-algorithm/src/v1.rs
Comment (rymnc) [reply crates/fuel-gas-price-algorithm/src/v1.rs:355]: if we just get finalized data we dont need to worry about reorgs too much

**Review by rafal-ch — DISMISSED** [review]

File: crates/fuel-gas-price-algorithm/src/v1.rs
Comment (rafal-ch) [inline crates/fuel-gas-price-algorithm/src/v1.rs:547]: I'm thinking about the following scenario, not sure if this is an issue or not:
1. we have 3 blocks in `self.unrecorded_blocks`
2. we `remove()` 2 of them, but we get an error on 3rd one
3. now, the `unrecorded_blocks_bytes` becomes out-of-sync

File: crates/fuel-gas-price-algorithm/src/v1.rs
Comment (MitchTurner) [reply crates/fuel-gas-price-algorithm/src/v1.rs:544]: I've been thinking about the (very unlikely) case that our gas price metadata is "ahead" of the chain on restart. This could in theory result in a bunch of the blocks being missing from the `unrecorded_blocks`. At that point we might have the da_committer reporting those blocks again. 

Since the precision of the algorithm isn't that important, I think the simplest solution would be to just ignore re-reported blocks.

In that case we don't want it throwing an error here. So I'm going to change this.

Comment (MitchTurner) [reply crates/fuel-gas-price-algorithm/src/v1.rs:544]: In terms of the algorithm, I think the best approach here is to ignore the batch. The problem is if it ever occurs, we will carry those `unrecorded_blocks` forever. It's fine from the algorithm's perspective because the error will become negligible pretty quickly.

Comment (MitchTurner) [reply crates/fuel-gas-price-algorithm/src/v1.rs:544]: Nevermind. We can both remove the blocks and ignore. I think that's the best.

**Review by rymnc — APPROVED** [review]

**Review by netrome — DISMISSED** [review]

File: crates/fuel-gas-price-algorithm/src/v1.rs
Comment (xgreenx) [inline crates/fuel-gas-price-algorithm/src/v1.rs:559]: It is not clear to me why missed bundles don't allow us to calculate the cost per byte. We know the total number of bytes and the cost of posting them. Why we can't use these values? What are we afraid of?

File: crates/services/gas_price_service/src/v1/metadata.rs
Comment (xgreenx) [inline crates/services/gas_price_service/src/v1/metadata.rs:12]: Do we have a task to remove `unrecorded_blocks` field from the `V1Metadata`?(We should use separate table to store them). Taking into account that committer posts bundles each hour, it can be a lot of storage space(each `V1Metadata` will have 3600 entries, and we will insert it into the storage for each block height).

File: crates/services/gas_price_service/src/v1/metadata.rs
Comment (xgreenx) [inline crates/services/gas_price_service/src/v1/metadata.rs:152]: Why `unrecorded_blocks_bytes` can't be a field of the metadata, instead of calculating it each time?

File: crates/fuel-gas-price-algorithm/src/v1.rs
Comment (xgreenx) [inline crates/fuel-gas-price-algorithm/src/v1.rs:579]: If block is not part of the `self.unrecorded_blocks`, it means we already posted it to the DA in the past.

We could either just ignore it or only increase the known DA cost on the amount of fee that we've paid to return this amount.

File: crates/fuel-gas-price-algorithm/src/v1.rs
Comment (MitchTurner) [reply crates/fuel-gas-price-algorithm/src/v1.rs:559]: > missed bundles

No. This is about missed blocks.

> We know the total number of bytes

Not really. We have the total bytes after compression. 

We could in theory go through each block in the bundle and query our api to find the original bytes for each block that's not listed in `unrecorded_blocks`, but this is an edge case that we probably will never hit and I think this solution works well enough since gas price is off-chain and a little error is fine.

File: crates/services/gas_price_service/src/v1/metadata.rs
Comment (MitchTurner) [reply crates/services/gas_price_service/src/v1/metadata.rs:12]: That's a good idea. I'll add a story. It doesn't need to be removed from `V1Metadata`, it just needs to be handled in the adapter.

Comment (MitchTurner) [reply crates/services/gas_price_service/src/v1/metadata.rs:12]: https://github.com/FuelLabs/fuel-core/issues/2454

File: crates/services/gas_price_service/src/v1/metadata.rs
Comment (MitchTurner) [reply crates/services/gas_price_service/src/v1/metadata.rs:152]: It only gets calculated on startup. I figured that would be better than adding another field to every metadata that is stored.

File: crates/fuel-gas-price-algorithm/src/v1.rs
Comment (MitchTurner) [reply crates/fuel-gas-price-algorithm/src/v1.rs:579]: > or only increase the known DA cost on the amount of fee that we've paid to return this amount.

I don't know what this means. I'm not sure what `fee` you're talking about

Ignoring makes the most sense to me but maybe I'm missing something.

File: crates/services/gas_price_service/src/v1/metadata.rs
Comment (xgreenx) [reply crates/services/gas_price_service/src/v1/metadata.rs:12]: I prefer the usage of the `unrecorded_blocks_bytes` field over fetching all unrecorded blocks on startup-up from the storage. Iterating over all elements of the storage only works on the `latest_view`. If you want to initiate the gas price service at a specific block height via `view_at,` then fetching will not work.

We don't have use cases right now where we need to support `view_at`, so it is not critical. But it would be nice if we had this option.

File: crates/services/gas_price_service/src/v1/metadata.rs
Comment (xgreenx) [reply crates/services/gas_price_service/src/v1/metadata.rs:152]: image

In the worst case, it is 481 MB per year. But since we use postcard, it will be much lower. Plus, we don't need to store data for the whole year. So I wouldn't concern about the storage=)

File: crates/fuel-gas-price-algorithm/src/v1.rs
Comment (xgreenx) [reply crates/fuel-gas-price-algorithm/src/v1.rs:579]: If the block committer submits two bundles(or more) for the same block height, it means that we pay a fee to DA twice(or more). Maybe we want to include the cost of posting the second bundle into the known DA cost.

File: crates/fuel-gas-price-algorithm/src/v1.rs
Comment (xgreenx) [reply crates/fuel-gas-price-algorithm/src/v1.rs:559]: Question to you: What is `new_cost_per_byte`? Is it the cost per 1 compressed byte submitted by the block committer to DA? Or is it the cost per 1 uncompressed byte for the Fuel block with the compression ratio included?

I would expect the first one to be used. But it seems you use the second in the code.

An example:

Block X
The Fuel Block is - 256KB.
The compressed Fuel Block is - 64KB.
Let's assume the bundle has 1 Fuel block(the block above).
The cost of posting this bundle is 1ETH.

I will expect that `new_cost_per_byte` is `1ETH / 64KB` because it is the actual cost to post bytes to the DA.
The current implementation will use `1ETH / 256KB`. So the `new_cost_per_byte` will be lower - cheaper.

Let's see what the cost will be in the next block, assuming we updated `new_cost_per_byte` by the rules above.

Block X + 1
The Fuel Block is - 256KB.
The projected cost with the current implementation is `256KB * 1 ETH / 256KB = 1 ETH`.
The projected cost I would expect - `256 KB * 1ETH / 64KB = 4 ETH`.

The compressed Fuel Block is - 256KB(all data is new, so compression is bad).
The current implementation will have `3 ETH` debt, assuming the same cost on DA. But if it increases, the debt will be more.

Comment (MitchTurner) [reply crates/fuel-gas-price-algorithm/src/v1.rs:559]: `new_cost_per_byte` is the cost per uncompressed bytes.

Absolutely, the compression ratio will be different from block-to-block. We can't get the fully compressed and zipped block, as that doesn't get produced until we actually commit (at which point projecting the cost isn't needed any more).

We could split the difference and get the pre-zipped, compressed block in theory. But I feel like for now, since the blobs include so many blocks, it averages out.

We could look into adding this improvement after we get DA gas costs running and before things get too congested.

File: crates/fuel-gas-price-algorithm/src/v1.rs
Comment (MitchTurner) [reply crates/fuel-gas-price-algorithm/src/v1.rs:579]: Ah ah. Yeah. That is possible in theory. But probably only when finalization fails on L1, right? In that case we don't pay the first fee, but only the second.

This is another "improvement" I've discussed. Where we track "committed but not finalized" blocks, not just `unrecorded_blocks`. I've considered this something we'd like to do after we get the DA gas cost live.

File: crates/services/gas_price_service/src/v1/metadata.rs
Comment (MitchTurner) [reply crates/services/gas_price_service/src/v1/metadata.rs:12]: Is this more about the next comment:

> Why unrecorded_blocks_bytes can't be a field of the metadata, instead of calculating it each time?

We still need the bytes for each unrecorded block, even if we store the total as well.

File: crates/fuel-gas-price-algorithm/src/v1.rs
Comment (xgreenx) [reply crates/fuel-gas-price-algorithm/src/v1.rs:559]: I think it would be simpler if `new_cost_per_byte` were just the last DA cost per 1 byte(regardless of whether it is compressed or uncompressed). In this case, in the example above, we will charge the user `4 ETH` based on the uncompressed data, assuming the worst-case without compression.

File: crates/services/gas_price_service/src/v1/metadata.rs
Comment (xgreenx) [reply crates/services/gas_price_service/src/v1/metadata.rs:12]: > We still need the bytes for each unrecorded block, even if we store the total as well.

Yes, we need it, but we can fetch this information during iteration over `heights` requesting `maybe_bytes` from the table in the database.

image

File: crates/fuel-gas-price-algorithm/src/v1.rs
Comment (xgreenx) [reply crates/fuel-gas-price-algorithm/src/v1.rs:579]: It is not additional "improvement", it is the case that you handle by this code=) Basically, it is only possible in the case described above. If `unrecorded_blocks` doesn't have some block, it is only possible if you already processed that block height in the past(so you had two or more bundles that covered this block height).

Based on your comments we can just do nothing for now, and cover this case later=)

File: crates/services/gas_price_service/src/v1/metadata.rs
Comment (MitchTurner) [reply crates/services/gas_price_service/src/v1/metadata.rs:152]: We can address it in https://github.com/FuelLabs/fuel-core/issues/2454

**Review by xgreenx — APPROVED** [review]

**Review by rymnc — APPROVED** [review]

**Pull request merged**
